include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+#define BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+
+namespace boost {
+namespace urls {
+
+class decoded_view::const_iterator
+{
+    char const* begin_{nullptr};
+    char const* pos_{nullptr};
+    char const* end_{nullptr};
+    bool plus_to_space_{true};
+
+    friend decoded_view;
+
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space) noexcept
+        : begin_(str)
+        , pos_(str)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    // end ctor
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
```

Comment:
how much is bidirectionality costing us and what are we getting out of it?

Reply:
I'm not sure I understand.

For most use cases, even in an algorithm that requires random access, I'm pretty sure iterating the view is going to be faster than allocating memory. In all other cases, we can still allocate.

Ergonomics apart, that seems like the trade-off.

Reply:
we don't need `begin_` if we are ForwardIterator only, so it is costing us at least one pointer. Is iterating backwards useful? What's the use-case?

Reply:
Oh... I thought you were talking about the cost in comparison to random access, which is what we had before.

Yes. Then the trade-off is 25% in the size of iterators against the use case of any algorithm that requires a reverse string or going forward and backwards in a string.

The 25% trade-off is easy to measure. The is such a general operation that it's not easy to list all use cases. These strings come from URIs and might contain all kinds of data: such as resource paths, things to be used in arbitrary database queries, and so on. 

I would imagine almost any use case for iterating strings that way would also be a use case for decoded strings. I'm not an expert but off the top of my head these would be are algorithms involving regular expressions, arithmetical operations on numbers represented as strings, finding rhymes in dictionaries, number base conversion, reverse binary searches to avoid common prefixes, identifying file extensions, and the classic palindrome function in programming interviews :).

Reply:
@pdimov ?

Reply:
or @glenfe ?

Reply:
I don't see the use-case for reverse iteration but some fat is always necessary to protect the internal organs I suppose.

Reply:
Mmm. You're right. I had a brainfart. Forward iterators should be fine (just not input only iterators).

Reply:
Actually, I just remembered a huge use case for reverse decoded iterators: `operator<(url_view, url_view)` in this very library. The algorithm path comparison in linear time needs to happen in reverse. In fact, the current implementation can be made much smaller now that we have the view.